File: Frontend/Web/src/app/services/notification.service.ts
import { Injectable, inject, signal, computed } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable, interval, switchMap, tap, catchError, of } from 'rxjs';

export interface Notification {
  id: number;
  title: string;
  message: string;
  type: string;
  is_read: boolean;
  pfe_listing_id: number | null;
  application_id: number | null;
  created_at: string;
}

export interface UnreadCountResponse {
  count: number;
}

@Injectable({
  providedIn: 'root'
})
export class NotificationService {
  private http = inject(HttpClient);
  private readonly API_URL = 'http://localhost:8000/api/notifications';
  private readonly TOKEN_KEY = 'pfe_match_token';

  // Signals for reactive state
  private notificationsSignal = signal<Notification[]>([]);
  private unreadCountSignal = signal<number>(0);
  private loadingSignal = signal<boolean>(false);

  // Public computed values
  notifications = this.notificationsSignal.asReadonly();
  unreadCount = this.unreadCountSignal.asReadonly();
  loading = this.loadingSignal.asReadonly();
  
  hasUnread = computed(() => this.unreadCountSignal() > 0);

  private getAuthHeaders(): HttpHeaders {
    const token = localStorage.getItem(this.TOKEN_KEY);
    let headers = new HttpHeaders({
      'Content-Type': 'application/json'
    });
    if (token) {
      headers = headers.set('Authorization', `Bearer ${token}`);
    }
    return headers;
  }

  /**
   * Fetch all notifications for the current user
   */
  getNotifications(unreadOnly: boolean = false): Observable<Notification[]> {
    this.loadingSignal.set(true);
    const url = unreadOnly ? `${this.API_URL}?unread_only=true` : this.API_URL;
    
    return this.http.get<Notification[]>(url, { headers: this.getAuthHeaders() }).pipe(
      tap(notifications => {
        this.notificationsSignal.set(notifications);
        this.loadingSignal.set(false);
      }),
      catchError(error => {
        console.error('Error fetching notifications:', error);
        this.loadingSignal.set(false);
        return of([]);
      })
    );
  }

  /**
   * Get unread notification count
   */
  getUnreadCount(): Observable<UnreadCountResponse> {
    return this.http.get<UnreadCountResponse>(`${this.API_URL}/unread-count`, { headers: this.getAuthHeaders() }).pipe(
      tap(response => {
        this.unreadCountSignal.set(response.count);
      }),
      catchError(error => {
        console.error('Error fetching unread count:', error);
        return of({ count: 0 });
      })
    );
  }

  /**
   * Mark a specific notification as read
   */
  markAsRead(notificationId: number): Observable<any> {
    return this.http.post(`${this.API_URL}/${notificationId}/read`, {}, { headers: this.getAuthHeaders() }).pipe(
      tap(() => {
        // Update local state
        const notifications = this.notificationsSignal();
        const updated = notifications.map(n => 
          n.id === notificationId ? { ...n, is_read: true } : n
        );
        this.notificationsSignal.set(updated);
        
        // Decrease unread count
        const currentCount = this.unreadCountSignal();
        if (currentCount > 0) {
          this.unreadCountSignal.set(currentCount - 1);
        }
      }),
      catchError(error => {
        console.error('Error marking notification as read:', error);
        return of(null);
      })
    );
  }

  /**
   * Mark all notifications as read
   */
  markAllAsRead(): Observable<any> {
    return this.http.post(`${this.API_URL}/read-all`, {}, { headers: this.getAuthHeaders() }).pipe(
      tap(() => {
        // Update local state
        const notifications = this.notificationsSignal();
        const updated = notifications.map(n => ({ ...n, is_read: true }));
        this.notificationsSignal.set(updated);
        this.unreadCountSignal.set(0);
      }),
      catchError(error => {
        console.error('Error marking all notifications as read:', error);
        return of(null);
      })
    );
  }

  /**
   * Start polling for new notifications
   */
  startPolling(intervalMs: number = 30000): Observable<UnreadCountResponse> {
    return interval(intervalMs).pipe(
      switchMap(() => this.getUnreadCount())
    );
  }

  /**
   * Refresh notifications (manual refresh)
   */
  refresh(): void {
    this.getNotifications().subscribe();
    this.getUnreadCount().subscribe();
  }

  /**
   * Get time ago string for notification
   */
  getTimeAgo(dateString: string): string {
    const date = new Date(dateString);
    const now = new Date();
    const seconds = Math.floor((now.getTime() - date.getTime()) / 1000);

    if (seconds < 60) return 'Just now';
    if (seconds < 3600) return `${Math.floor(seconds / 60)}m ago`;
    if (seconds < 86400) return `${Math.floor(seconds / 3600)}h ago`;
    if (seconds < 604800) return `${Math.floor(seconds / 86400)}d ago`;
    
    return date.toLocaleDateString();
  }
}

File: Frontend/mobile/lib/Services/notification_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../core/config/api_config.dart';
import '../models/notification.dart';
import 'token_service.dart';

class NotificationService {
  Future<Map<String, String>> _getHeaders() async {
    final token = await TokenService.getToken();
    return {
      'Content-Type': 'application/json',
      'Authorization': 'Bearer $token',
    };
  }

  /// Get all notifications for the current user
  Future<List<Notification>> getNotifications({bool unreadOnly = false}) async {
    try {
      final headers = await _getHeaders();
      final url = unreadOnly
          ? '${ApiConfig.baseUrl}/api/notifications?unread_only=true'
          : '${ApiConfig.baseUrl}/api/notifications';

      final response = await http.get(
        Uri.parse(url),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final List<dynamic> data = json.decode(response.body);
        return data.map((json) => Notification.fromJson(json)).toList();
      } else if (response.statusCode == 401) {
        throw Exception('Unauthorized. Please login again.');
      } else {
        throw Exception('Failed to load notifications: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching notifications: $e');
    }
  }

  /// Get unread notification count
  Future<int> getUnreadCount() async {
    try {
      final headers = await _getHeaders();
      final response = await http.get(
        Uri.parse('${ApiConfig.baseUrl}/api/notifications/unread-count'),
        headers: headers,
      );

      if (response.statusCode == 200) {
        final data = json.decode(response.body);
        return data['count'] as int;
      } else if (response.statusCode == 401) {
        throw Exception('Unauthorized. Please login again.');
      } else {
        throw Exception('Failed to load unread count: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error fetching unread count: $e');
    }
  }

  /// Mark a specific notification as read
  Future<void> markAsRead(int notificationId) async {
    try {
      final headers = await _getHeaders();
      final response = await http.post(
        Uri.parse('${ApiConfig.baseUrl}/api/notifications/$notificationId/read'),
        headers: headers,
        body: json.encode({}),
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to mark notification as read: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error marking notification as read: $e');
    }
  }

  /// Mark all notifications as read
  Future<void> markAllAsRead() async {
    try {
      final headers = await _getHeaders();
      final response = await http.post(
        Uri.parse('${ApiConfig.baseUrl}/api/notifications/read-all'),
        headers: headers,
        body: json.encode({}),
      );

      if (response.statusCode != 200) {
        throw Exception('Failed to mark all notifications as read: ${response.statusCode}');
      }
    } catch (e) {
      throw Exception('Error marking all notifications as read: $e');
    }
  }

  /// Get time ago string for notification
  String getTimeAgo(DateTime dateTime) {
    final now = DateTime.now();
    final difference = now.difference(dateTime);

    if (difference.inSeconds < 60) {
      return 'Just now';
    } else if (difference.inMinutes < 60) {
      return '${difference.inMinutes}m ago';
    } else if (difference.inHours < 24) {
      return '${difference.inHours}h ago';
    } else if (difference.inDays < 7) {
      return '${difference.inDays}d ago';
    } else {
      return '${dateTime.day}/${dateTime.month}/${dateTime.year}';
    }
  }
}
